_posts/2019-05-29-wtf-are-streams.md

```diff
@@ -0,0 +1,277 @@
+using System.Text;
+
+namespace StreamTime
+{
+    public class FileTheStream
+    {
+
+        public static void Main()
+        {
+            const string path = @"/Users/jgarcia/Desktop/example.txt";
+
+            //Create the file.
+            using (FileStream fs = File.Create(path))
+            {
+                AddText(fs, "This is some text");
+                AddText(fs, "This is some more text,");
+                AddText(fs, "\r\nand this is on a new line");
+                AddText(fs, "\r\n\r\nThe following is a subset of characters:\r\n");
+
+                for (var i = 1; i < 120; i++)
+                {
+                    AddText(fs, Convert.ToChar(i).ToString());
+                }
+            }
+
+            //Open the stream and read it back.
+            using (FileStream fs = File.OpenRead(path))
+            {
+                var b = new byte[1024];
+                var temp = new UnicodeEncoding();
+                while (fs.Read(b,0,b.Length) > 0)
+                {
+                    Console.WriteLine(temp.GetString(b));
+                }
+            }
+        }
+
+        private static void AddText(Stream fs, string value)
+        {
+            var info = new UnicodeEncoding().GetBytes(value);
+            fs.Write(info, 0, info.Length);
+        }
+    }
+}
+```
+
+As you can see, code-wise, we're just opening a stream, reading, writing, and closing it. Well, you might be wondering
+where the closing is happening. Take into account that in C#, whenever we use the `using` keyword with a resource,
+once it's finished using it, it closes the resource – Just like Java's *try with resources*. But let's talk further
+about the opening and closing of streams.
+
+### Why do streams have to be opened or closed?
+
+We mentioned that every time we open a stream the OS needs to dedicate resources, but we never talked about which,
+or how. Most of the times, it depends on the nature of the stream which we have opened – it's not the same to open
+a socket than to open a file, etc. For the time being, we can concentrate on the files.
+
+Whenever a new file stream is open, the OS dedicates a file descriptor, commonly known as a file handle, to the
+application. A file descriptor basically is a number that uniquely identifies an open file in a computer's OS – it
+describes a data source, and how it can be accessed. File descriptors point towards the kernel's global file table,
+which contains information such as the offset and the access restrictions of the stream.
+
+As you can imagine, file descriptors isn't like memory – if you don't return them to the OS the situation can grow
+ugly fast. It's just a matter of time until the application crashes in a long running application. This is usually
+called a file-handle leak. In Windows computers, one of the consequences is usually when you try to delete a file
+but it says that it's being used by another program. The resource hasn't been freed properly.
```

Comment:
In Windows computers, when you try to delete a file, usually one of the consequences is that it says it's being used by another program.